extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../../../src/3p-analytics-common';
+import {throttle} from '../../../src/utils/rate-limit';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.CrossDomainIframeMessageQueue';
+
+/** @private @const {number} */
+const MESSAGE_THROTTLE_TIME_ = 100;
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ * @abstract
+ */
+class AbstractAmpAnalytics3pMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates cross-frame communication
+   * @param {!string} messageType The type that will be used to
+   * contain the queued data (see 3p-analytics-common)
+   */
+  constructor(win, iframeMessagingClient, messageType) {
+    /** @type {boolean} */
+    this.isReady_ = false;
+
+    /** @type {!Window} */
+    this.win_ = win;
+
+    /** @type {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @type {string} */
+    this.messageType_ = messageType;
+
+    /** @type {!Object<string,!string|!Array<string>>} */
+    this.store_ = {};
```

Comment:
Can we make this variable name more meaningful?  Maybe creativeToPendingMessages?